homeassistant/components/dormakaba_dkey/__init__.py

```diff
@@ -0,0 +1,125 @@
+from py_dormakaba_dkey.models import AssociationData
+
+from homeassistant.components import bluetooth
+from homeassistant.components.bluetooth.match import ADDRESS, BluetoothCallbackMatcher
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_ADDRESS, EVENT_HOMEASSISTANT_STOP, Platform
+from homeassistant.core import Event, HomeAssistant, callback
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import CONF_ASSOCIATION_DATA, DEVICE_TIMEOUT, DOMAIN, UPDATE_SECONDS
+from .models import DormakabaDkeyData
+
+PLATFORMS: list[Platform] = [Platform.LOCK]
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Dormakaba dKey from a config entry."""
+    address: str = entry.data[CONF_ADDRESS]
+    ble_device = bluetooth.async_ble_device_from_address(hass, address.upper(), True)
+    if not ble_device:
+        raise ConfigEntryNotReady(f"Could not find dKey device with address {address}")
+
+    lock = DKEYLock(ble_device)
+    lock.set_association_data(
+        AssociationData.from_json(entry.data[CONF_ASSOCIATION_DATA])
+    )
+
+    @callback
+    def _async_update_ble(
+        service_info: bluetooth.BluetoothServiceInfoBleak,
+        change: bluetooth.BluetoothChange,
+    ) -> None:
+        """Update from a ble callback."""
+        lock.set_ble_device_and_advertisement_data(
+            service_info.device, service_info.advertisement
+        )
+
+    entry.async_on_unload(
+        bluetooth.async_register_callback(
+            hass,
+            _async_update_ble,
+            BluetoothCallbackMatcher({ADDRESS: address}),
+            bluetooth.BluetoothScanningMode.PASSIVE,
+        )
+    )
+
+    async def _async_update() -> None:
+        """Update the device state."""
+        try:
+            await lock.update()
+            await lock.disconnect()
+        except DKEY_EXCEPTIONS as ex:
+            raise UpdateFailed(str(ex)) from ex
+
+    startup_event = asyncio.Event()
+    cancel_first_update = lock.register_callback(lambda *_: startup_event.set())
+    coordinator = DataUpdateCoordinator(
+        hass,
+        _LOGGER,
+        name=lock.name,
+        update_method=_async_update,
+        update_interval=timedelta(seconds=UPDATE_SECONDS),
+    )
+
+    try:
+        await coordinator.async_config_entry_first_refresh()
+    except ConfigEntryNotReady:
+        cancel_first_update()
+        raise
+
+    try:
+        async with async_timeout.timeout(DEVICE_TIMEOUT):
+            await startup_event.wait()
+    except asyncio.TimeoutError as ex:
+        raise ConfigEntryNotReady(
+            "Unable to communicate with the device; "
+            f"Try moving the Bluetooth adapter closer to {lock.name}"
+        ) from ex
+    finally:
+        cancel_first_update()
```

Comment:
Do you need a separate `startup_event` here since you are using a `DataUpdateCoordinator`?

Reply:
Hmm, I was inspired by `led_ble` which also has the event even though it's using a `DataUpdateCoordinator`.

Reply:
The event is not needed and has been removed 👍